### PR TITLE
DebiasedLasso implementation.

### PR DESCRIPTION
**Pull request by moprescu**
* Added DebiasedLasso algorithm as a Lasso extension
* Moved all the lasso extensions to a `sklearn_extensions` file
* Fixed some bugs and wrote more tests

### CODE REVIEW

File: econml/sklearn_extensions.py [hunk 1]
```diff
@@ -0,0 +1,975 @@
...
+    n_iter_ : int | array-like, shape (n_targets,)
+        number of iterations run by the coordinate descent solver to reach
+        the specified tolerance.
+    """
+
+    def __init__(self, alpha=1.0, fit_intercept=True,
+                 precompute=False, copy_X=True, max_iter=1000,
+                 tol=1e-4, warm_start=False, positive=False,
+                 random_state=None, selection='cyclic'):
+        super(WeightedLasso, self).__init__(
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
        super().__init__(
```

File: econml/sklearn_extensions.py [hunk 2]
```diff
@@ -0,0 +1,975 @@
...
+
+    n_iter_ : int | array-like, shape (n_targets,)
+        number of iterations run by the coordinate descent solver to reach
+        the specified tolerance.
+    """
+
+    def __init__(self, alpha=1.0, fit_intercept=True, normalize=False,
+                 copy_X=True, max_iter=1000, tol=1e-4, warm_start=False,
+                 random_state=None, selection='cyclic'):
+        super(WeightedMultiTaskLasso, self).__init__(
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
        super().__init__(
```

File: econml/sklearn_extensions.py [hunk 3]
```diff
@@ -0,0 +1,975 @@
...
+        If None, the random number generator is the RandomState instance used
+        by `np.random`. Used when ``shuffle`` == True.
+    """
+
+    def __init__(self, n_splits=3, n_trials=10, shuffle=False, random_state=None):
+        self.n_splits = n_splits
+        self.shuffle = shuffle
+        self.n_trials = n_trials
+        self.random_state = random_state
+        return
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
```

File: econml/sklearn_extensions.py [hunk 4]
```diff
@@ -0,0 +1,975 @@
...
+
+        sample_weight : numpy array of shape [n_samples]
+                        Individual weights for each sample.
+                        The weights will be normalized internally.
+        """
+        _fit_weighted_linear_model(self, WeightedMultiTaskLasso, X, y, sample_weight)
+        return self
+
+
+class WeightedKFold(object):
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: We more frequently do

```suggestion
class WeightedKFold:
```

File: econml/sklearn_extensions.py [hunk 5]
```diff
@@ -0,0 +1,975 @@
...
+    intercept_std_err_ : float
+        Estimated standard error intercept (see 'intercept_' attribute).
+
+    """
+
+    def __init__(self, alpha='auto', fit_intercept=True,
+                 precompute=False, copy_X=True, max_iter=1000,
+                 tol=1e-4, warm_start=False, positive=False,
+                 random_state=None, selection='cyclic'):
+        super(DebiasedLasso, self).__init__(
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
        super().__init__(
```

File: econml/sklearn_extensions.py [hunk 6]
```diff
@@ -0,0 +1,975 @@
...
+                np.matmul(np.matmul(self._X_offset, self._coef_variance), self._X_offset) +
+                self._mean_error_variance
+            )
+        else:
+            self.intercept_std_err_ = 0
+
+        # Set intercept
+        self._set_intercept(X_offset, y_offset, X_scale)
+        # Return alpha to 'auto' state
+        if self.selected_alpha_ is not None:
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: Is there a risk that an exception will be thrown in the interim, leaving the state incorrect for a subsequent call to fit?  If so, might be better to put this within an `except` block.

File: econml/sklearn_extensions.py [hunk 7]
```diff
@@ -0,0 +1,975 @@
...
+            self.selected_alpha_ = self.alpha
+        else:
+            # Warn about consistency
+            warnings.warn("Setting a suboptimal alpha can lead to miscalibrated confidence intervals. "
+                          "We recommend setting alpha='auto' for optimality.")
+
+        # Convert X, y into numpy arrays
+        X, y = check_X_y(X, y, y_numeric=True, multi_output=False)
+        # Fit weighted lasso with user input
+        super(DebiasedLasso, self).fit(X, y, sample_weight, check_input)
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
        super().fit(X, y, sample_weight, check_input)
```

File: econml/sklearn_extensions.py [hunk 8]
```diff
@@ -0,0 +1,975 @@
...
+            X, y, fit_intercept=self.fit_intercept, normalize=False,
+            copy=self.copy_X, check_input=check_input, sample_weight=sample_weight, return_mean=True)
+
+        # Calculate quantities that will be used later on. Account for centered data
+        y_pred = self.predict(X) - self.intercept_
+        self._theta_hat = self._get_theta_hat(X, sample_weight)
+        self._X_offset = X_offset
+
+        # Calculate coefficient and error variance
+        num_nonzero_coefs = np.sum(self.coef_ != 0)
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
        num_nonzero_coefs = np.count_nonzero(self.coef_)
```

File: econml/sklearn_extensions.py [hunk 9]
```diff
@@ -0,0 +1,975 @@
...
+        # Add coefficient correction
+        coef_correction = self._get_coef_correction(
+            X, y, y_pred, sample_weight, self._theta_hat)
+        self.coef_ += coef_correction
+
+        # Set coefficients and intercept standard errors
+        self.coef_std_err_ = np.sqrt(np.diag(self._coef_variance))
+        if self.fit_intercept:
+            self.intercept_std_err_ = np.sqrt(
+                np.matmul(np.matmul(self._X_offset, self._coef_variance), self._X_offset) +
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
                self._X_offset @ self._coef_variance @ self._X_offset +
```

File: econml/sklearn_extensions.py [hunk 10]
```diff
@@ -0,0 +1,975 @@
...
+        Returns
+        -------
+        (y_lower, y_upper) : tuple of arrays, shape (n_samples, )
+            Returns lower and upper interval endpoints.
+        """
+        y_pred = self.predict(X)
+        y_lower = np.empty(y_pred.shape)
+        y_upper = np.empty(y_pred.shape)
+        # Note that in the case of no intercept, X_offset is 0
+        X = X - self._X_offset
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: Perhaps remove the comment and just be explicit:

```suggestion
        if self.fit_intercept:
            X = X - self._X_offset
```

File: econml/sklearn_extensions.py [hunk 11]
```diff
@@ -0,0 +1,975 @@
...
+        y_lower = y_pred + \
+            np.apply_along_axis(lambda s: norm.ppf(
+                lower / 100, scale=s), 0, sd_pred)
+        y_upper = y_pred + \
+            np.apply_along_axis(lambda s: norm.ppf(
+                upper / 100, scale=s), 0, sd_pred)
+        return y_lower, y_upper
+
+    def _get_coef_correction(self, X, y, y_pred, sample_weight, theta_hat):
+        # Assumes flattened y
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: Do we have tests that assert that we do the right thing when inputs are of different shapes (not specifically for this method, but in general)?

File: econml/sklearn_extensions.py [hunk 12]
```diff
@@ -0,0 +1,975 @@
...
+            coefs[i] = local_wlasso.coef_
+            # Weighted tau
+            if sample_weight is not None:
+                y_weighted = y * sample_weight / np.sum(sample_weight)
+            else:
+                y_weighted = y / n_samples
+            tausq[i] = np.dot(y - local_wlasso.predict(X_reduced), y_weighted)
+        # Compute C_hat
+        C_hat = np.diag(np.ones(n_features))
+        C_hat[0][1:] = - coefs[0]
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: I think for unary `-` juxtaposition looks better.

```suggestion
        C_hat[0][1:] = -coefs[0]
```

File: econml/sklearn_extensions.py [hunk 13]
```diff
@@ -0,0 +1,975 @@
...
+        for i in range(1, n_features):
+            C_hat[i][:i] = - coefs[i][:i]
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
            C_hat[i][:i] = -coefs[i][:i]
```

File: econml/sklearn_extensions.py [hunk 14]
```diff
@@ -0,0 +1,975 @@
...
+            C_hat[i][i + 1:] = - coefs[i][i:]
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
            C_hat[i][i + 1:] = -coefs[i][i:]
```

File: econml/sklearn_extensions.py [hunk 15]
```diff
@@ -0,0 +1,975 @@
...
+        theta_hat = np.matmul(np.diag(1 / tausq), C_hat)
+        return theta_hat
+
+    def _get_unscaled_coef_var(self, X, theta_hat, sample_weight):
+        if sample_weight is not None:
+            weights_mat = np.diag(sample_weight / np.sum(sample_weight))
+            sigma = np.matmul(X.T, np.matmul(weights_mat, X))
+        else:
+            sigma = np.matmul(X.T, X) / X.shape[0]
+        _unscaled_coef_var = np.matmul(
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]: Replace `np.matmul(np.matmul(X, Y), Z)` with equivalent `X @ Y @ Z`

I'd have provided this in a suggestion, but GitHub doesn't support multi-line suggestions yet.

File: econml/sklearn_extensions.py [hunk 16]
```diff
@@ -0,0 +1,975 @@
...
+            C_hat[i][:i] = - coefs[i][:i]
+            C_hat[i][i + 1:] = - coefs[i][i:]
+        # Compute theta_hat
+        theta_hat = np.matmul(np.diag(1 / tausq), C_hat)
+        return theta_hat
+
+    def _get_unscaled_coef_var(self, X, theta_hat, sample_weight):
+        if sample_weight is not None:
+            weights_mat = np.diag(sample_weight / np.sum(sample_weight))
+            sigma = np.matmul(X.T, np.matmul(weights_mat, X))
```
Comment (kbattocchi) [inline econml/sklearn_extensions.py]:
```suggestion
            sigma = X.T @ weights_mat @ X
```

File: econml/sklearn_extensions.py (hunk 9 above)
Comment (kbattocchi) [reply econml/sklearn_extensions.py]: Should one of those `self._X_offset`s be transposed?

File: econml/sklearn_extensions.py [hunk 17]
```diff
@@ -0,0 +1,975 @@
...
+# Licensed under the MIT License.
+
+"""Collection of scikit-learn extensions."""
+
+# TODO: file docs
+# TODO: move lasso extensions here
+# TODO: move debiased lasso here
+# TODO: run old lasso tests
+# TODO: debiased lasso tests
+# TODO: debiased lasso docs
```
[the remark this answers is not in the document]
Comment (moprescu) [reply econml/sklearn_extensions.py]: Oh, all these are done. I forgot to remove the TODOs. Doing that now!

File: econml/sklearn_extensions.py (hunk 15 above)
Comment (moprescu) [reply econml/sklearn_extensions.py]: @kbattocchi This is super cool, I didn't know you could do this. Replacing this is all my code. Thanks for the detailed comments on the PR, I'm learning some new stuff :)

File: econml/sklearn_extensions.py [hunk 18]
```diff
@@ -0,0 +1,975 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (moprescu) [reply econml/sklearn_extensions.py]: I did some restructuring based on Vasilis's comment. We can do further reorganizing later.

File: econml/sklearn_extensions.py [hunk 19]
```diff
@@ -0,0 +1,975 @@
...
+from scipy.stats import norm
+from sklearn.linear_model import LassoCV, MultiTaskLassoCV, Lasso, MultiTaskLasso
+from sklearn.model_selection import KFold, StratifiedKFold
+from sklearn.model_selection import GridSearchCV
+from sklearn.model_selection._split import _CVIterableWrapper, CV_WARNING
+from sklearn.utils import check_array, check_X_y
+from sklearn.utils.multiclass import type_of_target
+
+
+def _fit_weighted_linear_model(self, class_name, X, y, sample_weight, check_input=None):
```
[the remark this answers is not in the document]
Comment (moprescu) [reply econml/sklearn_extensions.py]: Done!

File: econml/tests/test_lasso.py [hunk 20]
```diff
@@ -203,56 +207,131 @@ def test_multiple_outputs_no_weights_cv(self):
...
+        self._check_debiased_coefs(TestLassoExtensions.X, TestLassoExtensions.y_simple + TestLassoExtensions.intercept,
+                                   sample_weight=None,
+                                   expected_coefs=TestLassoExtensions.coefs1, expected_intercept=TestLassoExtensions.intercept)
+        # --> Check 5-95 CI coverage for unit vectors
+        self._check_debiased_CI(TestLassoExtensions.X, TestLassoExtensions.y_simple + TestLassoExtensions.intercept,
+                                sample_weight=None,
+                                expected_coefs=TestLassoExtensions.coefs1, expected_intercept=TestLassoExtensions.intercept)
+        # Test DebiasedLasso with weights for one DGP
+        # Define weights
+        sample_weight = np.concatenate((np.ones(TestLassoExtensions.n_samples // 2),
```
[the remark this answers is not in the document]
Comment (moprescu) [reply econml/tests/test_lasso.py]: Done!

File: econml/sklearn_extensions.py [hunk 21]
```diff
@@ -0,0 +1,975 @@
...
+        weight_mat = np.diag(sqrt_weights)
+        X_weighted = np.matmul(weight_mat, X)
+        y_weighted = np.matmul(weight_mat, y)
+        fit_params['X'] = X_weighted
+        fit_params['y'] = y_weighted
+        if self.fit_intercept:
+            # Fit base class without intercept
+            self.fit_intercept = False
+            # Fit Lasso
+            super(class_name, self).fit(**fit_params)
```
[the remark this answers is not in the document]
Comment (moprescu) [reply econml/sklearn_extensions.py]: Great suggestion. Done!

**Pull request merged**